File: backend/app/schemas/bank_reconciliation.py
"""Pydantic schemas for bank reconciliation.

Shared contract for the ``/api/bank-reconciliation`` router. Money fields use
the ``MoneyAmount`` annotations (Decimal in Python, JSON number on the wire);
IDs are strings on the wire, parsed to UUID in the router. See
``backend/docs/bank-reconciliation.md``.
"""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field

from app.schemas.money import MoneyAmount, OptionalMoneyAmount


class BankTransactionResponse(BaseModel):
    id: str
    transaction_date: str
    posted_date: str | None
    amount: MoneyAmount
    currency: str
    description: str | None
    counterparty_name: str | None
    reference: str | None
    direction: str
    matched_payment_id: str | None
    matched_invoice_number: str | None = None
    match_method: str | None
    # A 0-100 confidence score, not a currency amount — plain float, not the
    # Decimal-exact MoneyAmount annotation.
    match_confidence: float | None = None
    matched_at: str | None
    # What the bank account was debited for the matched payment — the FX leg's
    # home-currency figure for an international payment, `Payment.amount`
    # otherwise (`services.bank_reconciliation.settlement_amount_and_currency`).
    # Lets a reviewer see both sides of a discrepancy without a second request.
    matched_payment_amount: OptionalMoneyAmount = None
    # The currency that settlement amount is denominated in. NULL when it can't
    # be established — which is also when the currency comparison is skipped.
    matched_payment_currency: str | None = None
    # The matched payment's own status — what makes a `status_conflict` row
    # readable ("the bank moved money against a payment we call `failed`").
    matched_payment_status: str | None = None
    # Signed gap between what the bank moved and what the payment authorises
    # (`services.bank_reconciliation.match_variance`). POSITIVE means the bank
    # took MORE than we authorised — the direction that matters for fraud.
    # NULL when the transaction is unmatched, or when the two sides are in
    # different currencies (subtracting across currencies isn't money).
    variance_amount: OptionalMoneyAmount = None
    # False for any discrepancy line (`amount_mismatch` / `currency_mismatch` /
    # `status_conflict`): linked to a payment, but it has NOT cleared. Mirrors
    # `services.bank_reconciliation.is_reconciled`.
    is_reconciled: bool = False


class BankStatementResponse(BaseModel):
    id: str
    account_identifier: str
    currency: str
    period_start: str
    period_end: str
    source_format: str
    file_key: str | None
    opening_balance: OptionalMoneyAmount = None
    closing_balance: OptionalMoneyAmount = None
    transaction_count: int
    # RECONCILED lines only — a discrepancy line is linked but not cleared,
    # and counting it here would report the discrepancy as resolved.
    matched_count: int
    # Debits linked to a payment whose amount disagrees. Computed on read from
    # the transactions themselves — no stored column, no migration.
    amount_mismatch_count: int = 0
    # Every linked-but-unreconciled line: the amount mismatches above PLUS
    # `currency_mismatch` and `status_conflict`. The single "something on this
    # statement needs a human" number.
    discrepancy_count: int = 0
    imported_at: str
    created_at: str
    # Transactions are included on the detail response only (list omits them).
    transactions: list[BankTransactionResponse] | None = None


class BankStatementListResponse(BaseModel):
    items: list[BankStatementResponse]
    total: int
    page: int
    page_size: int


class TransactionResolveRequest(BaseModel):
    """Manually set or clear a transaction's matched payment. ``None`` clears
    an existing (correct or incorrect) match back to unmatched.

    Typed ``uuid.UUID``, not ``str``: the router used to call ``uuid.UUID(...)``
    on the raw value with no handler, so a malformed id was a 500 rather than
    the 422 a validation failure owes the caller.
    """

    matched_payment_id: uuid.UUID | None = Field(default=None)


# --------------------------------------------------------------------------- #
# Outstanding items (`GET /api/bank-reconciliation/outstanding`)
# --------------------------------------------------------------------------- #


class UnclearedPaymentResponse(BaseModel):
    """A payment our books say went out that no bank line has claimed — the
    "outstanding cheques / payments in transit" side of a bank rec worksheet."""

    payment_id: str
    invoice_id: str
    invoice_number: str | None = None
    vendor_name: str | None = None
    amount: MoneyAmount
    # The currency `amount` is denominated in — the invoice's, since
    # `Payment.amount` is invoice-currency. `unmatched_debits` and
    # `discrepancies` both carried one; without it this bucket rendered in the
    # org's reporting currency and a multi-currency tenant saw the wrong symbol.
    currency: str | None = None
    method: str | None = None
    status: str
    # `submitted_at` → `completed_at` → `created_at`, the same fallback chain
    # the matcher's date window uses.
    sent_on: str | None = None
    days_outstanding: int | None = None


class UnmatchedDebitResponse(BaseModel):
    """A bank debit with no payment behind it — money left the account that we
    have no record of authorising."""

    transaction_id: str
    statement_id: str
    account_identifier: str
    transaction_date: str
    amount: MoneyAmount
    currency: str
    counterparty_name: str | None = None
    reference: str | None = None
    description: str | None = None


class DiscrepancyResponse(BaseModel):
    """A bank debit we identified as one of our payments that does NOT
    reconcile. The fraud/error bucket — `classification` says how it fails."""

    transaction_id: str
    statement_id: str
    account_identifier: str
    transaction_date: str
    # `amount_mismatch` | `currency_mismatch` | `status_conflict`
    # (`services.bank_reconciliation.UNRECONCILED_MATCH_METHODS`).
    classification: str
    bank_amount: MoneyAmount
    bank_currency: str
    # What the bank account was debited for the payment — the FX leg's
    # home-currency figure when it has one, `Payment.amount` otherwise.
    payment_amount: MoneyAmount
    payment_currency: str | None = None
    # Our books' own status for the payment. A `status_conflict` row is one
    # where this is NOT a dispatched status.
    payment_status: str | None = None
    # Positive = the bank took MORE than we authorised. Set for the
    # `amount_mismatch` class only: a cross-currency gap isn't money, and a
    # `status_conflict` agrees on the amount by definition.
    variance_amount: OptionalMoneyAmount = None
    payment_id: str
    invoice_number: str | None = None
    counterparty_name: str | None = None


class BankReconCurrencyTotal(BaseModel):
    """One currency's slice of an outstanding-items total.

    ``total`` is an EXACT decimal string, never a float — a money figure a user
    reads, matching every other whole-set rollup in this codebase."""

    currency: str
    total: str


class OutstandingItemsResponse(BaseModel):
    """Org-wide reconciliation state, computed on read across every imported
    statement — the period-close view the per-statement detail can't give."""

    as_of: str
    older_than_days: int
    uncleared_payments: list[UnclearedPaymentResponse]
    uncleared_count: int
    # Grouped per currency, NEVER a cross-currency SUM. `Payment.amount` is
    # invoice-currency, so totalling it across a multi-currency tenant produced
    # a figure denominated in nothing real — the same rule
    # `amount_mismatch_net_variance` below already states for subtraction.
    uncleared_totals: list[BankReconCurrencyTotal]
    unmatched_debits: list[UnmatchedDebitResponse]
    unmatched_debit_count: int
    # Same rule: a statement carries its own currency, and a tenant can import
    # statements for accounts in different ones.
    unmatched_debit_totals: list[BankReconCurrencyTotal]
    # Every identified-but-unreconciled line, whatever its class.
    discrepancies: list[DiscrepancyResponse]
    discrepancy_count: int
    # Signed sum of the AMOUNT-mismatch subset's variances. Positive = the bank
    # has taken more than we authorised in aggregate. Deliberately not summed
    # over the other classes: a cross-currency subtraction isn't money.
    amount_mismatch_net_variance: MoneyAmount

File: frontend/src/lib/types/bankReconciliation.ts
// Types for the Bank Reconciliation surface. Mirrors the JSON returned by the
// `/api/bank-reconciliation` endpoints (the Pydantic contract in
// `backend/app/schemas/bank_reconciliation.py`). Date fields are ISO strings
// (or null); money fields are `MoneyAmount`, never `number`.
//
// PII discipline mirrors the backend: `account_identifier` is whatever the
// importer typed to name the account (a nickname or a masked tail) — the full
// account / routing numbers never leave the bank's own file, and nothing here
// carries one. Nothing in this module widens what the API already exposes.
//
// Not entity-scoped: `BankStatement` / `BankTransaction` predate multi-entity
// and cover an org-wide bank account, so there is no entity filter here.

import type { BadgeTone } from '$lib/components/ui/Badge.svelte';
import type { MoneyAmount } from '$lib/utils/money';

// --- Match method ---------------------------------------------------------

/**
 * How a bank line came to be linked to one of our payments.
 *
 * The first four are *identity* claims of descending strength; the last three
 * are the discrepancy classes — the line is linked to a payment but does NOT
 * reconcile with it (`services/bank_reconciliation.UNRECONCILED_MATCH_METHODS`).
 */
export type MatchMethod =
	| 'provider_id'
	| 'amount_date'
	| 'fuzzy_vendor'
	| 'manual'
	| 'amount_mismatch'
	| 'currency_mismatch'
	| 'status_conflict';

/**
 * The three classes where the bank moved money the payment does not support.
 * Mirrors `UNRECONCILED_MATCH_METHODS` — a linked line carrying one of these
 * is deliberately NOT counted in `BankStatement.matched_count`.
 */
export const DISCREPANCY_METHODS: readonly string[] = [
	'amount_mismatch',
	'currency_mismatch',
	'status_conflict'
];

/** Is this match method one of the three discrepancy classes? */
export function isDiscrepancyMethod(method: string | null | undefined): boolean {
	return method !== null && method !== undefined && DISCREPANCY_METHODS.includes(method);
}

// --- Match state — the judgment a row is asking the reader to make ---------

/**
 * What a reviewer is actually looking at on one transaction row.
 *
 * Derived rather than stored, because the API answers the question across
 * four fields (`direction`, `matched_payment_id`, `match_method`,
 * `match_confidence`) and a row that renders only one of them tells a
 * half-truth. In particular a 50–70 fuzzy vendor-name hit is a *suggestion* a
 * human still owes a decision on — rendering it the same green as a reference
 * match would be the UI asserting a fact the matcher never claimed.
 *
 *  - `credit`       — not a payment we made; the matcher skips these entirely.
 *  - `unmatched`    — a debit with no payment behind it (money left the
 *                     account we have no record of authorising).
 *  - `discrepancy`  — linked to a payment, but it does not reconcile.
 *  - `confirmed`    — certain identity: an exact reference hit, or a human's
 *                     own resolve (both land at confidence 100).
 *  - `probable`     — a single amount + date candidate in the window (80–99).
 *  - `suggested`    — vendor-name similarity only (50–79), or a linked row
 *                     carrying no confidence at all. Needs a human.
 */
export type MatchState =
	| 'credit'
	| 'unmatched'
	| 'discrepancy'
	| 'confirmed'
	| 'probable'
	| 'suggested';

/** The minimal row shape {@link transactionMatchState} reads. */
export interface MatchStateInput {
	direction: string;
	matched_payment_id: string | null;
	match_method: string | null;
	match_confidence: number | null;
	is_reconciled: boolean;
}

/**
 * Collapse a transaction's four match fields into the one judgment the row
 * renders. Pure — no clock, no fetch — so it is unit-tested directly.
 *
 * Order matters. `is_reconciled` is the backend's own predicate
 * (`services/bank_reconciliation.is_reconciled`), so a linked line it calls
 * unreconciled is a discrepancy even if `match_method` is a value this
 * frontend has never heard of: an unknown method must degrade to "a human
 * needs to look", never to a clean green tick.
 */
export function transactionMatchState(tx: MatchStateInput): MatchState {
	if (tx.direction !== 'debit') return 'credit';
	if (!tx.matched_payment_id) return 'unmatched';
	if (isDiscrepancyMethod(tx.match_method) || !tx.is_reconciled) return 'discrepancy';
	const confidence = tx.match_confidence;
	// A linked, reconciled row with no confidence figure is not evidence of
	// certainty — it is the absence of evidence. Treat it as a suggestion.
	if (confidence === null || confidence === undefined) return 'suggested';
	if (confidence >= 100) return 'confirmed';
	if (confidence >= 80) return 'probable';
	return 'suggested';
}

/**
 * States where a human still owes a decision — the rows the page pulls
 * forward and the ones that get the prominent resolve affordance.
 */
export function needsHumanDecision(state: MatchState): boolean {
	return state === 'discrepancy' || state === 'suggested' || state === 'unmatched';
}

/**
 * Badge tone per match state, so the statement table and the transaction rows
 * can't tint the same state two different shades.
 *
 * `probable` is `accent` rather than `success` on purpose: an amount + date
 * coincidence in a ±5-day window is a strong identity claim, not a confirmed
 * one, and green reads as "signed off".
 */
export const MATCH_STATE_TONES: Record<MatchState, BadgeTone> = {
	credit: 'neutral',
	unmatched: 'warning',
	discrepancy: 'danger',
	confirmed: 'success',
	probable: 'accent',
	suggested: 'warning'
};

// --- Response shapes ------------------------------------------------------

export interface BankTransaction {
	id: string;
	transaction_date: string;
	posted_date: string | null;
	amount: MoneyAmount;
	currency: string;
	description: string | null;
	counterparty_name: string | null;
	reference: string | null;
	/** `debit` | `credit`. Only a debit can clear a payment. */
	direction: string;
	matched_payment_id: string | null;
	matched_invoice_number: string | null;
	match_method: string | null;
	/** A 0–100 identity score, NOT a currency amount — see `MatchState`. */
	match_confidence: number | null;
	matched_at: string | null;
	/** What the account was debited for the matched payment (the FX leg's
	 *  home-currency figure for an international payment). */
	matched_payment_amount: MoneyAmount;
	matched_payment_currency: string | null;
	matched_payment_status: string | null;
	/** Signed gap: POSITIVE means the bank took MORE than we authorised.
	 *  Null when unmatched or when the two sides are in different currencies. */
	variance_amount: MoneyAmount;
	/** False for every discrepancy class — linked, but it has not cleared. */
	is_reconciled: boolean;
}

export interface BankStatement {
	id: string;
	account_identifier: string;
	currency: string;
	period_start: string;
	period_end: string;
	source_format: string;
	file_key: string | null;
	opening_balance: MoneyAmount;
	closing_balance: MoneyAmount;
	/** Row counts, not money. */
	transaction_count: number;
	/** RECONCILED lines only — a discrepancy line is linked but not cleared. */
	matched_count: number;
	amount_mismatch_count: number;
	/** Every linked-but-unreconciled line: the single "needs a human" number. */
	discrepancy_count: number;
	imported_at: string;
	created_at: string;
	/** Present on the detail response only; the list omits them. */
	transactions: BankTransaction[] | null;
}

// --- Outstanding items (`GET /outstanding`) --------------------------------

export interface UnclearedPayment {
	payment_id: string;
	invoice_id: string;
	invoice_number: string | null;
	vendor_name: string | null;
	amount: MoneyAmount;
	/** The currency `amount` is denominated in (the invoice's). */
	currency: string | null;
	method: string | null;
	status: string;
	sent_on: string | null;
	/** Age in days — a duration, not money. */
	days_outstanding: number | null;
}

export interface UnmatchedDebit {
	transaction_id: string;
	statement_id: string;
	account_identifier: string;
	transaction_date: string;
	amount: MoneyAmount;
	currency: string;
	counterparty_name: string | null;
	reference: string | null;
	description: string | null;
}

export interface Discrepancy {
	transaction_id: string;
	statement_id: string;
	account_identifier: string;
	transaction_date: string;
	/** One of `DISCREPANCY_METHODS`. */
	classification: string;
	bank_amount: MoneyAmount;
	bank_currency: string;
	payment_amount: MoneyAmount;
	payment_currency: string | null;
	payment_status: string | null;
	/** Set for the amount class only — a cross-currency gap isn't money. */
	variance_amount: MoneyAmount;
	payment_id: string;
	invoice_number: string | null;
	counterparty_name: string | null;
}

/** One currency's slice of an outstanding-items total. `total` is an EXACT
 *  decimal string, matching every other whole-set rollup in this app. */
export interface BankReconCurrencyTotal {
	currency: string;
	total: string;
}

export interface OutstandingItems {
	as_of: string;
	older_than_days: number;
	uncleared_payments: UnclearedPayment[];
	uncleared_count: number;
	/** Grouped per currency as exact decimal strings — never one blended sum.
	 *  `Payment.amount` is invoice-currency, so a cross-currency total is
	 *  denominated in nothing real. Render via `formatCurrencyTotals`. */
	uncleared_totals: BankReconCurrencyTotal[];
	unmatched_debits: UnmatchedDebit[];
	unmatched_debit_count: number;
	/** Same rule — a statement carries its own currency. */
	unmatched_debit_totals: BankReconCurrencyTotal[];
	discrepancies: Discrepancy[];
	discrepancy_count: number;
	/** Signed sum of the amount-mismatch subset. Positive = the bank has taken
	 *  more than we authorised in aggregate. */
	amount_mismatch_net_variance: MoneyAmount;
}

/**
 * Did `/outstanding` cap a bucket's rows?
 *
 * Every count and total on that response covers the FULL set while the row
 * list is capped at `?limit`, so a page that filters the rows client-side has
 * to say when it is looking at less than the count above it claims. Pure.
 */
export function isTruncated(shown: number, total: number): boolean {
	return shown < total;
}

/** The paginated statement-list envelope.
 *
 *  `total` is a pagination ROW COUNT, not money — recorded as such in
 *  `moneyTypeAudit.test.ts`'s `JUDGED_NOT_MONEY`, so the ratchet does not
 *  re-raise it every round. */
export interface BankStatementListResponse {
	items: BankStatement[];
	total: number;
	page: number;
	page_size: number;
}
